File: insecureapplication-go/gallery-idp/db/oauth.go
package db

import (
	"database/sql"
)

type AuthCode struct {
	Code        string
	ClientID    string
	UserID      string
	RedirectURI string
	Scope       string
	CreatedAt   string
}

type AccessToken struct {
	Token     string
	ClientID  string
	UserID    sql.NullString
	Scope     sql.NullString
	ExpiresIn int64
	CreatedAt string
}

type RefreshToken struct {
	Token     string
	ClientID  string
	UserID    string
	Scope     sql.NullString
	ExpiresIn int64
	CreatedAt string
}

func (d *DB) CreateAuthCode(code, clientID, userID, redirectURI, scope string) error {
	_, err := d.Exec(`INSERT INTO authorization_codes (code, client_id, user_id, redirect_uri, scope) VALUES (?, ?, ?, ?, ?)`,
		code, clientID, userID, redirectURI, scope)
	return err
}

// GetAuthCode looks a code up by code alone -- deliberately no client_id
// filter (see doc/OAuth2_PoC_Verification_Report.md PoC4: authorization code
// not bound to client). Rows are never deleted after being exchanged (PoC5:
// codes are reusable), and there is no expiry check (no such column exists).
func (d *DB) GetAuthCode(code string) (*AuthCode, error) {
	row := d.QueryRow(`SELECT code, client_id, user_id, redirect_uri, scope, created_at FROM authorization_codes WHERE code = ?`, code)
	c := &AuthCode{}
	var redirectURI, scope sql.NullString
	err := row.Scan(&c.Code, &c.ClientID, &c.UserID, &redirectURI, &scope, &c.CreatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	c.RedirectURI = redirectURI.String
	c.Scope = scope.String
	return c, nil
}

func (d *DB) CreateAccessToken(token, clientID, userID, scope string) error {
	_, err := d.Exec(`INSERT INTO access_tokens (token, client_id, user_id, scope) VALUES (?, ?, ?, ?)`,
		token, clientID, nullableString(userID), scope)
	return err
}

func (d *DB) GetAccessToken(token string) (*AccessToken, error) {
	row := d.QueryRow(`SELECT token, client_id, user_id, scope, expires_in, created_at FROM access_tokens WHERE token = ?`, token)
	t := &AccessToken{}
	err := row.Scan(&t.Token, &t.ClientID, &t.UserID, &t.Scope, &t.ExpiresIn, &t.CreatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return t, nil
}

func (d *DB) CreateRefreshToken(token, clientID, userID, scope string) error {
	_, err := d.Exec(`INSERT INTO refresh_tokens (token, client_id, user_id, scope) VALUES (?, ?, ?, ?)`,
		token, clientID, userID, scope)
	return err
}

// GetRefreshToken: deliberately no client_id filter (PoC6: refresh token not
// bound to client -- any client that obtains a valid refresh token, by any
// means, can redeem it for a fresh access token inheriting its scope).
func (d *DB) GetRefreshToken(token string) (*RefreshToken, error) {
	row := d.QueryRow(`SELECT token, client_id, user_id, scope, expires_in, created_at FROM refresh_tokens WHERE token = ?`, token)
	t := &RefreshToken{}
	err := row.Scan(&t.Token, &t.ClientID, &t.UserID, &t.Scope, &t.ExpiresIn, &t.CreatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return t, nil
}

func nullableString(s string) any {
	if s == "" {
		return nil
	}
	return s
}

File: insecureapplication/gallery/db/oauth.js
const {getDb} = require('./db');
const {orNull} = require('./id');
const config = require('../config/config');

// grantcode()/granttoken()/etc. generate codes and tokens as JS numbers
// (Math.floor(Math.random() * ...)), matching the original's
// Math.floor(...) values that Mongoose auto-stringified when saved to a
// String-typed field. node:sqlite's parameter binding isn't as forgiving:
// an integer-valued JS `number` can get bound as SQLite REAL rather than
// INTEGER, and a REAL coerced into one of these TEXT PRIMARY KEY columns
// stringifies with a trailing ".0" (e.g. "86907.0") instead of matching the
// plain-string "86907" that arrives when a client posts code=86907 in a
// token exchange. Coercing to String() at the DB boundary keeps every
// code/token column holding, and looked up by, the same plain-integer text
// regardless of whether the caller passed a JS number or string.
function id(value) {
  return value == null ? value : String(value);
}

// vulnerability: no expiry column, and rows are never deleted after use
// (see doc/OAuth2_PoC_Verification_Report.md PoC3/PoC5).
function createAuthCode(code, clientId, userId, redirectUri, scope) {
  getDb().prepare(
      `INSERT INTO authorization_codes (code, client_id, user_id, redirect_uri, scope) VALUES (?, ?, ?, ?, ?)`,
  ).run(id(code), clientId, userId, orNull(redirectUri), orNull(scope));
}

// GetAuthCode looks a code up by code alone -- deliberately no client_id
// filter (PoC4: authorization code not bound to client).
function getAuthCode(code) {
  return getDb().prepare(`SELECT * FROM authorization_codes WHERE code = ?`).get(id(code));
}

function createAccessToken(token, clientId, userId, scope, expiresIn) {
  getDb().prepare(
      `INSERT INTO access_tokens (token, client_id, user_id, scope, expires_in) VALUES (?, ?, ?, ?, ?)`,
  ).run(id(token), clientId, userId || null, orNull(scope), expiresIn == null ? config.token.access.expires_in : expiresIn);
}

function getAccessToken(token) {
  return getDb().prepare(`SELECT * FROM access_tokens WHERE token = ?`).get(id(token));
}

function createRefreshToken(token, clientId, userId, scope, expiresIn) {
  getDb().prepare(
      `INSERT INTO refresh_tokens (token, client_id, user_id, scope, expires_in) VALUES (?, ?, ?, ?, ?)`,
  ).run(id(token), clientId, userId, orNull(scope), expiresIn == null ? config.token.refresh.expires_in : expiresIn);
}

// GetRefreshToken: deliberately no client_id filter (PoC6: refresh token
// not bound to client).
function getRefreshToken(token) {
  return getDb().prepare(`SELECT * FROM refresh_tokens WHERE token = ?`).get(id(token));
}

// SQLite's datetime('now') format ("YYYY-MM-DD HH:MM:SS", UTC, no
// timezone marker) isn't directly parseable by `new Date()` as UTC in every
// JS engine, so this is shared by isExpired() and tokeninfo().
function parseCreatedAt(createdAt) {
  return new Date(createdAt.replace(' ', 'T') + 'Z').getTime() / 1000;
}

// Matches accesstoken.js/refreshtoken.js's isExpired(): actually called by
// middlewares/auth.js's BearerStrategy (unlike gallery-idp's Go port, where
// the equivalent check is implemented but deliberately never wired up --
// this Node app's original code really does enforce it).
function isExpired(token) {
  return Date.now() / 1000 > parseCreatedAt(token.created_at) + token.expires_in;
}

module.exports = {
  createAuthCode, getAuthCode,
  createAccessToken, getAccessToken,
  createRefreshToken, getRefreshToken,
  isExpired, parseCreatedAt,
};
